cache/Cache.js

```diff
@@ -0,0 +1,63 @@
+const EventEmitter = require('events')
+const _createRedisClient = require('./util/createRedisClient')
+const msgpack = require('msgpack-lite')
+
+// default entry expiration time in seconds
+const DEFAULT_EXPIRATION = 86400
+
+class Cache extends EventEmitter {
+  constructor ({ logger, nodes, defaultTtl, redisClientOptions }) {
+    super()
+    this._logger = logger
+    this._defaultTtl = defaultTtl || DEFAULT_EXPIRATION
+
+    this._onReady = () => this.emit('ready')
+    this._onError = (err) => this.emit('error', err)
+    this._onReconnecting = () => this.emit('reconnecting')
+    this._onNodeError = (err) => this.emit('node error', err)
+
+    const redisClient = this._redisClient =
+      _createRedisClient(nodes, redisClientOptions)
+
+    // forward events to clients so that they
+    // can handle situations appropriately
+    redisClient
+      .on('ready', this._onReady)
+      .on('reconnecting', this._onReconnecting)
+      .on('node error', this._onNodeError)
+      .on('error', this._onError)
+  }
+
+  async get (key) {
+    const encoded = await this._redisClient.getBuffer(key)
+    return encoded ? msgpack.decode(encoded) : encoded
+  }
+
+  async set (key, value, ttl) {
+    const data = msgpack.encode(value)
+    return this._redisClient.setex(key, ttl || this._defaultTtl, data)
+  }
+
+  async getEntity (key, ModelType) {
+    const data = await this.get(key)
+    return data ? ModelType.wrap(data) : data
```

Comment:
Should never happen, but technically if data is corrupted, `ModelType.wrap()` could have errors. We should add an array of errors as the second arg and log/throw if `errors.length`:

```js
async getEntity (key, ModelType) {
  const data = await this.get(key)
  if (!data) return data
  
  let errors = []
  data = ModelType.wrap(data, errors)
  
  if (errors.length) {
    logger.error(`Errors occurred while wrapping model: "${errors.join(',')}"`)
    return // If data is corrupt, we should return undefined
  }
  return data
}
```